### DISCUSSION TITLE
Switch to nvim-nio for async

### DISCUSSION THREAD

**Discussion by cseickel** [nvim-neo-tree/neo-tree.nvim#1342] [Internal Development]
https://github.com/nvim-neotest/nvim-nio

This was mentioned a few times by @pysan3 and I'd like to make a separate issue for it. I took a look at nio and I do really like it. Things that will be helpful for us:

1. It supports cancellation (this is huge)
2. It has built-in methods for lsp and some libuv fs actions (It might be a good idea to see about  adding directory scanning into nio itself if we use it)
3. Maybe it can be used to fix the hell that is testing asynchronous actions in neo-tree.
4. It is a really nice and simple api.

If they had support for running jobs like plenary, and if we switch to pathlib, then I don't think we need plenary at all anymore as a runtime requirement. Not that it really matters as everyone has plenary installed anyhow.

**Comment by pysan3** [comment]
2.
I've got that already in pathlib.nvim: https://github.com/pysan3/pathlib.nvim?tab=readme-ov-file#scan-directories
and it supports `depth` and `follow_symlinks`: https://pysan3.github.io/pathlib.nvim/doc/PathlibPath.html#PathlibPath.fs_iterdir
Example: https://github.com/pysan3/pathlib.nvim/blob/main/spec/measure_time_with_linux_source_code.lua
(Run it with `nvim --headless -c ':so ./spec/measure_time_with_linux_source_code.lua' -c ':noa qa!'` and `tail -f measure.log`)

3.
For tests, you might want to look into this: https://github.com/lewis6991/nvim-test
I haven't used it yet but I saw it being recommended in some places.

---

plenary.job-like feature is being worked on right now: https://github.com/nvim-neotest/nvim-nio/issues/5

**Reply by pysan3** [reply]
Quick update. `nio.process` is now merged and released as v1.5.0.
